zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,289 @@
+#![allow(dead_code, unused_variables)]
+use super::{Transaction, TransparentInput};
+use crate::{
+    parameters::ConsensusBranchId, serialization::ZcashSerialize, types::BlockHeight, Network,
+    NetworkUpgrade,
+};
+use blake2b_simd::Hash;
+use byteorder::{LittleEndian, WriteBytesExt};
+use std::io;
+
+const ZCASH_SIGHASH_PERSONALIZATION_PREFIX: &[u8; 12] = b"ZcashSigHash";
+const ZCASH_PREVOUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashPrevoutHash";
+const ZCASH_SEQUENCE_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSequencHash";
+const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashOutputsHash";
+const ZCASH_JOINSPLITS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashJSplitsHash";
+const ZCASH_SHIELDED_SPENDS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSSpendsHash";
+const ZCASH_SHIELDED_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSOutputHash";
+
+pub const SIGHASH_ALL: u32 = 1;
+const SIGHASH_NONE: u32 = 2;
+const SIGHASH_SINGLE: u32 = 3;
+const SIGHASH_MASK: u32 = 0x1f;
+const SIGHASH_ANYONECANPAY: u32 = 0x80;
+
+pub struct SigHasher<'a> {
+    pub(crate) trans: &'a Transaction,
+    pub(crate) hash_type: u32,
+    pub(crate) network: Network,
+    pub(crate) height: BlockHeight,
+}
+
+impl<'a> SigHasher<'a> {
+    pub fn sighash(self) -> Hash {
+        use NetworkUpgrade::*;
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(&self.personal())
+            .to_state();
+
+        match self.network_upgrade() {
+            Genesis => unimplemented!(),
+            BeforeOverwinter => unimplemented!(),
```

Comment:
Question: what happens if a malicious block with a BeforeOverwinter height tries to do a sighash?

Should we reject the block, rather than crashing?

Reply:
yea, I expect we'd probably want to make this return an `Option` if we have legit cases where we just wont hash a block, there isn't a zip for the sighash impl before overwinter and librustzcash doesn't even implement it so I decided not to start there.

Reply:
Let's change these to `unimplemented!("Zebra checkpoints on Sapling activation")` so that the message has context alerting us to a problem.  We should never be in a position where we're doing verification of pre-Sapling blocks, so if we find ourself here a program invariant has been violated and we should crash immediately.